[USDU-280] Reduced the chance of import timing out for complex meshes.

vickycl opened the pull request:
## Purpose of this PR

**Ticket/Jira #:** [USDU-280](https://jira.unity3d.com/browse/USDU-280) / #324

<!-- Description of feature/change. Links to screenshots, design docs, user docs, etc. Remember reviewers may be outside your team, and not know your feature/area that should be explained more. -->

For certain mesh setups (eg where st and normals attributes are of faceVarying interpolation) we have to unweld every face of the mesh to get consistent array sizes. During this sanitization we can timeout and report an error if it takes >1s.

- Increased the timeout on waiting for the read job from 1 second to 10 seconds
- Optimized the Triangulate() function to remove dynamically sized Lists, reducing allocations.

## Testing

**Functional Testing status:**

<!-- Explanation of what's tested, how tested and existing or new automation tests. Can include manual testing by self and/or QA. Specify test plans. Rarely acceptable to have no testing.  -->

Tested that the original timeout no longer reproduces.

**Performance Testing status:**

<!-- Could this PR affect performance? If so, what has been done to measure time taken / memory used etc? Can include new or existing performance tests. Also see [Ensuring Performance by Default](https://confluence.unity3d.com/display/DEV/Ensuring+Performance+by+Default). -->

This PR halved the time taken for triangulation in my local testing, profiled with Superluminal:

Before:
![image](https://user-images.githubusercontent.com/45685026/215528463-074ee974-2b02-444c-a9ca-92a91c9f4c8d.png)

After:
![image](https://user-images.githubusercontent.com/45685026/215528650-f9f41651-5440-4928-9dd3-f40238846d52.png)

## Overall Product Risks
<!-- See Testing Status, Complexity and Halo Effect for your PR](https://confluence.unity3d.com/display/DEV/Testing+Status%2C+Complexity+and+Halo+Effect+for+your+PR). -->

**Complexity:** Low
<!-- (Minimal / Low / Medium / High) -->

**Halo Effect:** Low
<!-- (Minimal / Low / Medium / High) -->

## Additional information

**Note to reviewers:**

<!-- Info per person for what to focus on, or historical info to understand who have previously reviewed and coverage. Help them get context. -->

@judubu please advise of any further situations I should test?

**Reminder:**
<!-- Things to remember to do before finalizing this PR. -->
- [x] Add entry in CHANGELOG.md _(if applicable)_

vickycl committed:
Reduced the chance of import timing out for complex meshes.

- Increased the timeout from 1 second to 10 seconds
- Optimized the Triangulate() function to remove dynamically sized Lists, reducing allocations

vickycl committed:
Update CHANGELOG.md

vickycl committed:
Fixes from testing

lee-aandrew commented on package/com.unity.formats.usd/Runtime/Scripts/IO/ReadJob.cs:
```diff
@@ -140,7 +140,7 @@ public bool MoveNext()
 
                 j++;
-                if (!m_ready.WaitOne(1000))
+                if (!m_ready.WaitOne(10000))
```
Just a question, is the increase in time out still intentional with the optimization?
If so, is there a need for a discussion on the right value for a time out, and if there is also a need for something like a "cancel import"  button during the import load time?

vickycl replied:
Yeah, it's still needed. I gambled with 10s being an 'acceptable' timeout given that the Triangulate code is twice as fast. Theoretically a file 20x the size of the repro case will hit this same thing, but given the specific conditions that have triggered it in this case (st and normals set to faceVarying) and the easy workaround for users encountering this in future (increase the timeout) I think this is acceptable _for now_. In future we should definitely look at a more thorough fix, and a cancel option would be very valid. @SimonBoorer mentioned the possibility of using Hydra's triangulate capability, I don't know if that would necessarily be faster but that's a little way off yet.

lee-aandrew replied:
Thank you for the explanation!

lee-aandrew commented on package/com.unity.formats.usd/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 - Disabled plugins on unsupported platforms.
 - Fixed a bug causing PointInstances to be duplicated.
 - Fixed a bug causing an ApplicationException "all sample times in times and baseTime must either all be numeric or all be default" to be thrown when refreshing a PointInstancer.
+- Optimized triangulation to reduce the chance of a "Timed out while waiting for thread" error on importing a complex mesh.
```
I did a comparison between the imports pre and post this change, and with my sample file, there was a reduction from 10 seconds to 5 seconds, which seems like a pretty big cut down in time and matches with Vicky's local testing

My test file has:
Batches: 42
Tris: 232.8M
Verts: 116.4M

judubu reviewed (approved):
Awesome gains. Thanks for optimizing.

vickycl committed:
Further triangulation optimization

As the new faces are created linearly, the face indeces in faceMapping are always consecutive. As we know this, we can just store the index of the first face created for each of the original faces, and use the originalFaceVertexCounts array to work out the rest.

michaeljblain commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -466,7 +466,12 @@ public static void ImportSkinning(string path, SanitizedMeshSample usdMesh, Mesh
                     {
-                        newFaceIndices.AddRange(usdMesh.faceMapping[faceIndex]);
+                        int firstFace = usdMesh.faceMapping[faceIndex];
+                        int faceCount = usdMesh.originalFaceVertexCounts[faceIndex] - 2;
```
Super minor: change variable name to faceVertexCount to align with originalFaceVertexCounts

vickycl replied:
It isn't actually the vertex count, it is the number of faces generated by the triangulation. So for example a quad face (vertexCount = 4) generates 2 triangle faces (vertexCount - 2). I'll change it to  triCount and add a comment :)

michaeljblain replied:
Ah yes triCount is perfect.

michaeljblain commented on package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs:
```diff
@@ -200,52 +201,71 @@ internal void Triangulate(bool changeHandedness)
-                var next = last + 1;
-                var t = 0;
-                for (; t < faceVertexCounts[i] - 2; t++)
+            if (changeHandedness)
```
Would be good to state which handedness we're changing from/to

vickycl replied:
I added a comment- does it make sense?

michaeljblain replied:
Looks good.

michaeljblain reviewed (changes requested):
Overall looks good. Some minor suggestions.

vickycl committed:
Improved naming and commenting from PR review.

vickycl committed:
Merge branch 'dev' into usdu-280/fix-import-timeout

vickycl committed:
Merge branch 'dev' into usdu-280/fix-import-timeout
